src/test/kotlin/org/jitsi/nlj/dtls/DtlsTest.kt

```diff
@@ -47,81 +45,69 @@ class DtlsTest : ShouldSpec() {
-                serverReceiver.processPacket(packetInfo)
-            }
-
-            override fun trace(f: () -> Unit) = f.invoke()
-        })
-
-        // We attach a consumer to each peer's receiver to consume the DTLS app packet
-        // messages
         val serverReceivedData = CompletableFuture<String>()
         val serverToClientMessage = "Goodbye, world"
-        serverReceiver.attach(object : ConsumerNode("server incoming app packets") {
-            override fun consume(packetInfo: PacketInfo) {
-                val packetData = ByteBuffer.wrap(packetInfo.packet.buffer, packetInfo.packet.offset, packetInfo.packet.length)
+        dtlsServer.incomingDataHandler = object : DtlsStack.IncomingDataHandler {
+            override fun dataReceived(data: ByteArray, off: Int, len: Int) {
+                val packetData = ByteBuffer.wrap(data, off, len)
                 val receivedStr = StandardCharsets.UTF_8.decode(packetData).toString()
                 debug("Server received message: '$receivedStr'")
                 serverReceivedData.complete(receivedStr)
-                serverSender.processPacket(PacketInfo(UnparsedPacket(serverToClientMessage.toByteArray())))
+                val serverToClientData = serverToClientMessage.toByteArray()
+
+                dtlsServer.sendApplicationData(serverToClientData, 0, serverToClientData.size)
             }
+        }
 
-            override fun trace(f: () -> Unit) = f.invoke()
-        })
+        // The DTLS client's send is wired directly to the DTLS server's receive
+        dtlsClient.outgoingDataHandler = object : DtlsStack.OutgoingDataHandler {
+            override fun sendData(data: ByteArray, off: Int, len: Int) {
+                pcapWriter?.processPacket(PacketInfo(UnparsedPacket(data, off, len)))
+                dtlsServer.processIncomingProtocolData(data, off, len)
+            }
+        }
 
         val clientReceivedData = CompletableFuture<String>()
-        clientReceiver.attach(object : ConsumerNode("client incoming app packets") {
-            override fun consume(packetInfo: PacketInfo) {
-                val packetData = ByteBuffer.wrap(packetInfo.packet.buffer, packetInfo.packet.offset, packetInfo.packet.length)
+        dtlsClient.incomingDataHandler = object : DtlsStack.IncomingDataHandler {
+            override fun dataReceived(data: ByteArray, off: Int, len: Int) {
+                val packetData = ByteBuffer.wrap(data, off, len)
                 val receivedStr = StandardCharsets.UTF_8.decode(packetData).toString()
                 debug("Client received message: '$receivedStr'")
                 clientReceivedData.complete(receivedStr)
             }
-
-            override fun trace(f: () -> Unit) = f.invoke()
-        })
+        }
 
         val serverThread = thread {
             debug("Server accepting")
             dtlsServer.start()
             debug("Server accepted connection")
         }
 
+        Thread.sleep(3000)
```

Comment:
I take it this is necessary to ensure the client starts after the server?

Reply:
Nope, I added this while doing some debugging at one point, it shouldn't be needed (since it wasn't there before).